### PR TITLE
Add support for casting to containers

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stacked on https://github.com/NVIDIA/Fuser/pull/531. This PR adds support to casting from `DynamicType` to containers and nested containers of concrete type. Also, this PR improves debugability by reporting the current dtype in the error message.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/dynamic_type.h [hunk 1]
```diff
@@ -214,6 +215,12 @@ struct DynamicType {
           return result;
         }(value)) {}
 
+  const std::type_info& type() const {
+    return std::visit(
```
Comment (naoyam) [inline csrc/dynamic_type.h:219]: C++ question: why this can't be just:

```
return typeid(value_);
```

**Review by naoyam — APPROVED** [review]
LGTM. Just left a minor question

File: csrc/dynamic_type.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/dynamic_type.h:219]: `return typeid(value_);` would return the `type_info` for the `std::variant<...>`, but I actually wanted the `type_info` for the actual dtype. For example, if the variant holds a `double`, then it should return the `type_info` of `double`.

Comment (zasdfgbnm) [reply csrc/dynamic_type.h:221]: Comment example added. Example error message:
```C++
using IntOrVec = DynamicType<Containers<std::vector>, int>;
IntOrVec v = 3;
(std::vector<double>)v;
// Cannot cast from i to St6vectorIdSaIdEE : incompatible type
```

**Pull request merged**